### PR TITLE
added bitand, bitor and tanimoto functions

### PR DESCRIPTION
**Pull request by robegan21**
Hello,
I've extended bitarray to include more efficient bitand, bitor (like bitdiff) and a highly optimized tanimoto function.  I've added documentation and unit tests.  Please consider pulling these changes into your repository.
Thanks,
Rob Egan

### CODE REVIEW

File: bitarray/_bitarray.c [hunk 1]
```diff
@@ -89,6 +89,79 @@ static PyTypeObject Bitarraytype;
...
+
+#define COUNT_BITS_ON(val, sum) \
+{\
+	_CBITS(val) \
+	sum = sum + _CBITS2(val); \
+}
+
+#define COUNT_BITS_ON_2_SUMS(val1, sum1, val2, sum2) \
+{\
+	_CBITSa(val1) _CBITSa(val2) \
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: Uh, a <tab> sneaked in! ;)

File: bitarray/_bitarray.c [hunk 2]
```diff
@@ -89,6 +89,79 @@ static PyTypeObject Bitarraytype;
...
+    _CBITSb(val1) _CBITSb(val2) \
+    _CBITSc(val1) _CBITSc(val2) \
+    sum1 = sum1 + _CBITS2(val1); \
+    sum2 = sum2 + _CBITS2(val2); \
+}
+
+#define COUNT_BITS_ON_2_WORDS(val1, val2, sum) \
+{\
+	_CBITS(val1) \
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: another tab

File: bitarray/_bitarray.c [hunk 3]
```diff
@@ -89,6 +89,79 @@ static PyTypeObject Bitarraytype;
...
+    _CBITS(val2) \
+    sum = sum + _CBITS2(val1) + _CBITS2(val2); \
+}
+
+#define COUNT_BITS_ON_4_WORDS(val1, val2, val3, val4, sum) \
+{\
+	_CBITS(val1) \
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: and some more.  I'll stop pointing to them individually now.

File: bitarray/_bitarray.c [hunk 4]
```diff
@@ -97,14 +170,32 @@ static PyTypeObject Bitarraytype;
...
     mask = BITMASK(self->endian, i);
     cp = self->ob_item + i / 8;
     if (bit)
         *cp |= mask;
     else
         *cp &= ~mask;
+    self->nbits_set = -1;
+}
+
+/* sets ususet bits to 0, i.e. the ones in the last byte (if any),
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: ususet -> unused?

File: bitarray/_bitarray.c [hunk 5]
```diff
@@ -131,6 +222,8 @@ resize(bitarrayobject *self, idx_t nbits)
 {
     Py_ssize_t newsize;
     size_t _new_size;       /* for allocation */
+//    void *newmem;
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: there are several commented out code snippets -- can those be cleaned out, or some documentation be added why they should be left in?

File: bitarray/_bitarray.c [hunk 6]
```diff
@@ -443,19 +548,59 @@ static int bitcount_lookup[256] = {
...
 }
+static idx_t
+count_old(bitarrayobject *self)
+{
+	setunused(self);
+	return self->nbits_set = count_range(self, 0, Py_SIZE(self));
+}
+
+static idx_t
+count_alt(bitarrayobject *self)
```
Comment (andre-merzky) [inline bitarray/_bitarray.c]: There are several `*_alt` and `*_old` methods in the code, which are also exposed in the API.  Could you clarify what those are?  Left for benchmarking purposes?

**Comment by diamondman** [comment]
Thanks @andre-merzky for putting work into reviewing the PR. I am setting up a fork on  Pypi and circleCI. lets see what we can get merged in!

**Comment by andre-merzky** [comment]
Hey @robegan21, 

this repo has been silent for quite a while, and we thus created a fork at https://github.com/diamondman/bitarray/.   

If you are interested, please feel free to transplant your pull request to the forked repo, we would be very happy to begin the code review and merge into bitarray before pushing out a new release.  If you do not have the time to do so, we would kindly ask your permission to do the PR transfer our-self.  If you could ping back in the next couple of days, one way or the other, that would be great.

Many thanks!

**Comment by ilanschnell** [comment]
I've added functions for efficiently counting and, or, xor in the new `bitarray.utils` module, which will be part of the 1.2.0 release.  The tanimoto functions seem too specific to be the bitarray library, sorry.

**Pull request closed without merging**